File: cpp/control_app.cpp
#include "control_app.h"
#include <QVBoxLayout>
#include <QGridLayout>
#include <QGroupBox>
#include <QMessageBox>
#include <QApplication>
#include <QScreen>
#include <QCloseEvent>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <unistd.h>
#include <ctime>
#include <chrono>
#include <iostream>

// 시스템 함수와 Qt 함수의 충돌을 피하기 위한 네임스페이스 사용
namespace sys {
    using ::close;
    using ::connect;
}

ControlApp::ControlApp(QWidget *parent)
    : QMainWindow(parent), isToggleOn(false), eventSent(false)
{
    setWindowTitle("Control Panel");
    
    // Initialize backends
    backends = {
        {
            "localhost",
            {9090, 9091},
            "Backend 1",
            false,
            {-1, -1}
        },
        {
            "localhost",
            {9092, 9093},
            "Backend 2",
            false,
            {-1, -1}
        }
    };
    
    setupUI();
    
    // Setup timers
    timer = new QTimer(this);
    connect(timer, &QTimer::timeout, this, &ControlApp::enableEventButton);
    
    statusTimer = new QTimer(this);
    connect(statusTimer, &QTimer::timeout, this, &ControlApp::connectToServer);
    statusTimer->start(1000);  // Check every 1 second
    
    eventBtn->setEnabled(true);
}

ControlApp::~ControlApp()
{
    // Clean up sockets
    for (auto& backend : backends) {
        for (int sock : backend.sockets) {
            if (sock != -1) {
                sys::close(sock);
            }
        }
    }
}

void ControlApp::setupUI()
{
    QWidget* centralWidget = new QWidget(this);
    setCentralWidget(centralWidget);
    QVBoxLayout* mainLayout = new QVBoxLayout(centralWidget);
    mainLayout->setSpacing(20);
    
    // Config group
    QGroupBox* configGroup = new QGroupBox("Backend Configuration", this);
    QGridLayout* configLayout = new QGridLayout;
    configLayout->setSpacing(10);
    
    for (size_t i = 0; i < backends.size(); ++i) {
        QLabel* ipLabel = new QLabel(QString::fromStdString(backends[i].name + " IP:"), this);
        ipLabel->setStyleSheet("font-size: 32px;");
        
        QLineEdit* ipInput = new QLineEdit(QString::fromStdString(backends[i].host), this);
        ipInput->setPlaceholderText("Enter IP address");
        ipInput->setStyleSheet("font-size: 32px; padding: 5px;");
        
        configLayout->addWidget(ipLabel, i, 0);
        configLayout->addWidget(ipInput, i, 1);
        ipInputs.push_back(ipInput);
        
        QString portText = QString("Ports: %1, %2")
            .arg(backends[i].ports[0])
            .arg(backends[i].ports[1]);
        QLabel* portLabel = new QLabel(portText, this);
        portLabel->setStyleSheet("font-size: 32px;");
        configLayout->addWidget(portLabel, i, 2);
    }
    
    applyBtn = new QPushButton("Apply Configuration", this);
    applyBtn->setMinimumSize(200, 50);
    applyBtn->setStyleSheet(
        "QPushButton {"
        "   font-size: 32px;"
        "   font-weight: bold;"
        "   padding: 5px;"
        "   background-color: #008CBA;"
        "   color: white;"
        "   border-radius: 5px;"
        "}"
        "QPushButton:hover {"
        "   background-color: #007399;"
        "}"
    );
    connect(applyBtn, &QPushButton::clicked, this, &ControlApp::applyConfiguration);
    configLayout->addWidget(applyBtn, backends.size(), 0, 1, 4);
    
    configGroup->setLayout(configLayout);
    mainLayout->addWidget(configGroup);
    
    // Control panel
    QGroupBox* controlGroup = new QGroupBox("Control Panel", this);
    QGridLayout* controlLayout = new QGridLayout;
    
    for (size_t i = 0; i < backends.size(); ++i) {
        QLabel* label = new QLabel(QString::fromStdString(backends[i].name + ": Not Connected"), this);
        label->setStyleSheet("color: red; font-size: 32px;");
        controlLayout->addWidget(label, 0, i, Qt::AlignCenter);
        statusLabels.push_back(label);
    }
    
    toggleBtn = new QPushButton("Start", this);
    toggleBtn->setMinimumSize(200, 50);
    toggleBtn->setStyleSheet(
        "QPushButton {"
        "   font-size: 32px;"
        "   font-weight: bold;"
        "   padding: 5px;"
        "   background-color: #4CAF50;"
        "   color: white;"
        "   border-radius: 5px;"
        "}"
        "QPushButton:hover {"
        "   background-color: #45a049;"
        "}"
    );
    
    eventBtn = new QPushButton("Send Event", this);
    eventBtn->setMinimumSize(200, 50);
    eventBtn->setStyleSheet(
        "QPushButton {"
        "   font-size: 32px;"
        "   font-weight: bold;"
        "   padding: 5px;"
        "   background-color: #008CBA;"
        "   color: white;"
        "   border-radius: 5px;"
        "}"
        "QPushButton:hover {"
        "   background-color: #007399;"
        "}"
        "QPushButton:disabled {"
        "   background-color: #cccccc;"
        "   color: #666666;"
        "}"
    );
    
    connect(toggleBtn, &QPushButton::clicked, this, &ControlApp::toggleAction);
    connect(eventBtn, &QPushButton::clicked, this, &ControlApp::sendEvent);
    
    controlLayout->addWidget(toggleBtn, 1, 0, 1, 2, Qt::AlignCenter);
    controlLayout->addWidget(eventBtn, 2, 0, 1, 2, Qt::AlignCenter);
    
    controlGroup->setLayout(controlLayout);
    mainLayout->addWidget(controlGroup);
    
    setMinimumSize(1200, 800);
    resize(1200, 800);
    centerWindow();
    
    setStyleSheet(
        "QGroupBox {"
        "   font-size: 32px;"
        "   font-weight: bold;"
        "   margin-top: 1ex;"
        "}"
        "QGroupBox::title {"
        "   subcontrol-origin: margin;"
        "   subcontrol-position: top center;"
        "   padding: 0 3px;"
        "}"
        "QLineEdit {"
        "   padding: 5px;"
        "   border: 1px solid #999;"
        "   border-radius: 3px;"
        "}"
    );
}

void ControlApp::connectToServer()
{
    bool allConnected = true;
    
    for (size_t i = 0; i < backends.size(); ++i) {
        auto& backend = backends[i];
        
        // First message exchange (1 -> 2)
        for (size_t j = 0; j < backend.ports.size(); ++j) {
            if (backend.sockets[j] != -1) continue;
            
            try {
                int sock = socket(AF_INET, SOCK_STREAM, 0);
                if (sock == -1) {
                    throw std::runtime_error("Socket creation failed");
                }
                
                struct timeval tv;
                tv.tv_sec = 0;
                tv.tv_usec = 500000;  // 0.5 seconds
                setsockopt(sock, SOL_SOCKET, SO_RCVTIMEO, &tv, sizeof(tv));
                
                struct sockaddr_in serverAddr;
                serverAddr.sin_family = AF_INET;
                serverAddr.sin_port = htons(backend.ports[j]);
                serverAddr.sin_addr.s_addr = inet_addr(backend.host.c_str());
                
                if (sys::connect(sock, (struct sockaddr*)&serverAddr, sizeof(serverAddr)) < 0) {
                    sys::close(sock);
                    throw std::runtime_error("Connection failed");
                }
                
                // Send first message and receive response
                // TODO: Implement protocol header message exchange
                
                backend.sockets[j] = sock;
                
            } catch (const std::exception& e) {
                std::cerr << "Error with " << backend.name << ":" << backend.ports[j] 
                         << ": " << e.what() << std::endl;
                allConnected = false;
                statusLabels[i]->setText(QString::fromStdString(backend.name + ": Not Connected"));
                statusLabels[i]->setStyleSheet("color: red; font-size: 32px;");
                continue;
            }
        }
        
        // Second message exchange (3 -> 4)
        if (std::all_of(backend.sockets.begin(), backend.sockets.end(), 
                       [](int sock) { return sock != -1; })) {
            try {
                for (size_t j = 0; j < backend.ports.size(); ++j) {
                    // TODO: Implement second message exchange
                }
                
                backend.ready = true;
                statusLabels[i]->setText(QString::fromStdString(backend.name + ": Connected"));
                statusLabels[i]->setStyleSheet("color: green; font-size: 32px;");
                
            } catch (const std::exception& e) {
                std::cerr << "Error with " << backend.name << " second message exchange: " 
                         << e.what() << std::endl;
                allConnected = false;
                statusLabels[i]->setText(QString::fromStdString(backend.name + ": Not Connected"));
                statusLabels[i]->setStyleSheet("color: red; font-size: 32px;");
                
                for (int sock : backend.sockets) {
                    if (sock != -1) sys::close(sock);
                }
                std::fill(backend.sockets.begin(), backend.sockets.end(), -1);
            }
        }
    }
    
    if (allConnected && std::all_of(backends.begin(), backends.end(), 
                                   [](const Backend& b) { return b.ready; })) {
        statusTimer->stop();
        eventSent = false;
        eventBtn->setEnabled(true);
        toggleBtn->setEnabled(true);
        std::cout << "All backends connected successfully" << std::endl;
    }
}

void ControlApp::applyConfiguration()
{
    for (size_t i = 0; i < backends.size(); ++i) {
        QString ip = ipInputs[i]->text().trimmed();
        if (ip.isEmpty()) {
            QMessageBox::warning(this, "Configuration Error", 
                               QString::fromStdString(backends[i].name + ": IP address cannot be empty"));
            return;
        }
        backends[i].host = ip.toStdString();
    }
    QMessageBox::information(this, "Success", "Configuration applied successfully");
}

void ControlApp::centerWindow()
{
    QScreen* screen = QGuiApplication::primaryScreen();
    QRect screenGeometry = screen->geometry();
    int x = (screenGeometry.width() - width()) / 2;
    int y = (screenGeometry.height() - height()) / 2;
    move(x, y);
}

void ControlApp::toggleAction()
{
    if (!isToggleOn) {
        auto [success, failedBackends] = sendTcpMessage("START");
        if (success) {
            toggleBtn->setText("End");
            toggleBtn->setStyleSheet(
                "QPushButton {"
                "   font-size: 32px;"
                "   font-weight: bold;"
                "   padding: 5px;"
                "   background-color: #ff9999;"
                "   color: white;"
                "   border-radius: 5px;"
                "}"
                "QPushButton:hover {"
                "   background-color: #ff8080;"
                "}"
            );
            isToggleOn = true;
            eventBtn->setEnabled(false);
        } else {
            if (!failedBackends.empty()) {
                std::string failureMessage = "CONNECTION_FAIL:";
                for (const auto& name : failedBackends) {
                    failureMessage += name + ",";
                }
                failureMessage.pop_back();  // Remove last comma
                sendTcpMessage(failureMessage);
            }
            toggleBtn->setText("Start");
            toggleBtn->setStyleSheet(
                "QPushButton {"
                "   font-size: 32px;"
                "   font-weight: bold;"
                "   padding: 5px;"
                "   background-color: #4CAF50;"
                "   color: white;"
                "   border-radius: 5px;"
                "}"
                "QPushButton:hover {"
                "   background-color: #45a049;"
                "}"
            );
            isToggleOn = false;
            eventBtn->setEnabled(true);
        }
    } else {
        auto [success, _] = sendTcpMessage("END");
        if (success) {
            toggleBtn->setText("Start");
            toggleBtn->setStyleSheet(
                "QPushButton {"
                "   font-size: 32px;"
                "   font-weight: bold;"
                "   padding: 5px;"
                "   background-color: #4CAF50;"
                "   color: white;"
                "   border-radius: 5px;"
                "}"
                "QPushButton:hover {"
                "   background-color: #45a049;"
                "}"
            );
            isToggleOn = false;
            eventBtn->setEnabled(true);
        }
    }
}

void ControlApp::sendEvent()
{
    // TODO: Implement event sending logic
}

void ControlApp::enableEventButton()
{
    eventBtn->setEnabled(true);
    timer->stop();
}

void ControlApp::closeEvent(QCloseEvent *event)
{
    for (auto& backend : backends) {
        for (int sock : backend.sockets) {
            if (sock != -1) {
                sys::close(sock);
            }
        }
    }
    event->accept();
}

std::pair<bool, std::vector<std::string>> ControlApp::sendTcpMessage(const std::string& message)
{
    // TODO: Implement TCP message sending logic
    return {true, {}};
} 

File: python/control_app.py
import sys
from PyQt5.QtWidgets import (QApplication, QMainWindow, QPushButton, QVBoxLayout, 
                           QWidget, QMessageBox, QLabel, QGridLayout, QLineEdit,
                           QGroupBox, QFormLayout)
from PyQt5.QtCore import Qt, QTimer
import socket
import threading
import time
from tcp_common import ProtocolHeader
import numpy as np
import boost.python as bp
from data_record_config_msg import DataRecordConfigMsgHandler, DataRecordConfigMsg, Header


class ControlApp(QMainWindow):
    def __init__(self):
        super().__init__()
        self.setWindowTitle("Control Panel")
        
        # TCP/IP settings for two backends
        self.backends = [
            {"host": "localhost", "ports": [9090, 9091], "name": "Backend 1", 
             "ready": False, "sockets": [None, None]},
            {"host": "localhost", "ports": [9092, 9093], "name": "Backend 2",
             "ready": False, "sockets": [None, None]}
        ]
        self.is_toggle_on = False
        self.event_sent = False
        self.message_counter = 0
        
        # Create central widget and layout
        central_widget = QWidget()
        self.setCentralWidget(central_widget)
        main_layout = QVBoxLayout(central_widget)
        main_layout.setSpacing(20)  # Add more space between groups
        
        # Create IP configuration group
        config_group = QGroupBox("Backend Configuration")
        config_layout = QGridLayout()
        config_layout.setSpacing(10)  # Add space between elements
        
        # Create input fields for each backend
        self.ip_inputs = []
        
        for i, backend in enumerate(self.backends):
            # IP input
            ip_label = QLabel(f"{backend['name']} IP:")
            ip_label.setStyleSheet("font-size: 32px;")
            ip_input = QLineEdit(backend['host'])
            ip_input.setPlaceholderText("Enter IP address")
            ip_input.setStyleSheet("font-size: 32px; padding: 5px;")
            config_layout.addWidget(ip_label, i, 0)
            config_layout.addWidget(ip_input, i, 1)
            self.ip_inputs.append(ip_input)
            
            # Port labels (fixed ports)
            port_label = QLabel(f"Ports: {backend['ports'][0]}, {backend['ports'][1]}")
            port_label.setStyleSheet("font-size: 32px;")
            config_layout.addWidget(port_label, i, 2)
        
        # Add apply button
        self.apply_btn = QPushButton("Apply Configuration")
        self.apply_btn.setMinimumSize(200, 50)
        self.apply_btn.clicked.connect(self.apply_configuration)
        self.apply_btn.setStyleSheet("""
            QPushButton {
                font-size: 32px;
                font-weight: bold;
                padding: 5px;
                background-color: #008CBA;
                color: white;
                border-radius: 5px;
            }
            QPushButton:hover {
                background-color: #007399;
            }
        """)
        config_layout.addWidget(self.apply_btn, len(self.backends), 0, 1, 4)
        
        config_group.setLayout(config_layout)
        main_layout.addWidget(config_group)
        
        # Create control group
        control_group = QGroupBox("Control Panel")
        control_layout = QGridLayout()
        
        # Create status labels
        self.status_labels = []
        for i, backend in enumerate(self.backends):
            label = QLabel(f"{backend['name']}: Not Connected")
            label.setStyleSheet("color: red; font-size: 32px;")
            control_layout.addWidget(label, 0, i, alignment=Qt.AlignCenter)
            self.status_labels.append(label)
        
        # Create buttons with larger font
        self.toggle_btn = QPushButton("Start")
        self.toggle_btn.setMinimumSize(200, 50)
        self.toggle_btn.setStyleSheet("""
            QPushButton {
                font-size: 32px;
                font-weight: bold;
                padding: 5px;
                background-color: #4CAF50;
                color: white;
                border-radius: 5px;
            }
            QPushButton:hover {
                background-color: #45a049;
            }
        """)
        self.toggle_btn.clicked.connect(self.toggle_action)
        
        self.event_btn = QPushButton("Send Event")
        self.event_btn.setMinimumSize(200, 50)
        self.event_btn.setStyleSheet("""
            QPushButton {
                font-size: 32px;
                font-weight: bold;
                padding: 5px;
                background-color: #008CBA;
                color: white;
                border-radius: 5px;
            }
            QPushButton:hover {
                background-color: #007399;
            }
            QPushButton:disabled {
                background-color: #cccccc;
                color: #666666;
            }
        """)
        self.event_btn.clicked.connect(self.send_event)
        
        # Add buttons to layout
        control_layout.addWidget(self.toggle_btn, 1, 0, 1, 2, alignment=Qt.AlignCenter)
        control_layout.addWidget(self.event_btn, 2, 0, 1, 2, alignment=Qt.AlignCenter)
        
        control_group.setLayout(control_layout)
        main_layout.addWidget(control_group)
        
        # Set initial window size and position
        self.setMinimumSize(1200, 800)  # Increased window size to accommodate larger fonts
        self.resize(1200, 800)
        self.center_window()
        
        # Create timer for re-enabling event button
        self.timer = QTimer()
        self.timer.timeout.connect(self.enable_event_button)
        
        # Create timer for checking backend status
        self.status_timer = QTimer()
        self.status_timer.timeout.connect(self.connect_to_server)
        self.status_timer.start(1000)  # Check every 1 second
        
        # Set window style
        self.setStyleSheet("""
            QGroupBox {
                font-size: 32px;
                font-weight: bold;
                margin-top: 1ex;
            }
            QGroupBox::title {
                subcontrol-origin: margin;
                subcontrol-position: top center;
                padding: 0 3px;
            }
            QLineEdit {
                padding: 5px;
                border: 1px solid #999;
                border-radius: 3px;
            }
        """)
        
        self.event_btn.setEnabled(True)  # Enable event button in initial state
        
    def connect_to_server(self):
        all_connected = True
        message_handler = DataRecordConfigMsgHandler()
        
        # 각 백엔드에 대해
        for i, backend in enumerate(self.backends):
            prev_counter = self.message_counter
            try:
                s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
                s.settimeout(0.5)
                s.connect((backend["host"], backend["ports"][0]))

                # First message (MessageType 1)
                self.message_counter += 1
                first_msg = Header(time.time_ns(), 1, self.message_counter, 0)
                s.sendall(first_msg.tobytes())
                response_data = s.recv(21)
                response = Header.from_bytes(response_data)
                if response.message_type != 2:
                    raise Exception(f"Expected MessageType 2, got {response.message_type}")
                
                # Second message (MessageType 3)
                self.message_counter += 1
                second_msg = Header(time.time_ns(), 3, self.message_counter, 0)
                s.sendall(second_msg.tobytes())
                response_data = s.recv(21)
                response = Header.from_bytes(response_data)
                if response.message_type != 4:
                    raise Exception(f"Expected MessageType 2, got {response.message_type}")

                backend["ready"] = True
                self.status_labels[i].setText(f"{backend['name']}: Connected")
                self.status_labels[i].setStyleSheet("color: green; font-size: 32px;")
                
            except Exception as e:
                print(f"Error with {backend['name']}:{backend['ports'][0]}: {e}")
                all_connected = False
                self.status_labels[i].setText(f"{backend['name']}: Not Connected")
                self.status_labels[i].setStyleSheet("color: red; font-size: 32px;")
                if 's' in locals():
                    s.close()
                self.message_counter = prev_counter
                continue
            
        if all_connected and all(backend["ready"] for backend in self.backends):
            self.status_timer.stop()
            self.event_sent = False
            self.event_btn.setEnabled(True)
            self.toggle_btn.setEnabled(True)
            print("All backends connected successfully")
            return True
        
        return False
        
    def apply_configuration(self):
        for i, backend in enumerate(self.backends):
            try:
                # Validate IP
                ip = self.ip_inputs[i].text().strip()
                if not ip:
                    raise ValueError(f"{backend['name']}: IP address cannot be empty")
                
                # Update backend configuration
                self.backends[i]['host'] = ip
                
            except ValueError as e:
                QMessageBox.warning(self, "Configuration Error", str(e))
                return
        
        QMessageBox.information(self, "Success", "Configuration applied successfully")

    def center_window(self):
        screen = QApplication.desktop().screenGeometry()
        size = self.geometry()
        self.move(
            (screen.width() - size.width()) // 2,
            (screen.height() - size.height()) // 2
        )
        
    def set_message_content(self, message):
        message.BodyLength = 0
        return message

    def send_tcp_message(self, start=True):
        message_handler = DataRecordConfigMsgHandler()
        
        if start:
            self.message_counter += 1
            config_msg = DataRecordConfigMsg(
                header=Header(time.time_ns(), 19, self.message_counter, 0),
                logging_directory_path="",
                logging_mode=np.uint32(0),
                history_time=np.uint32(0),
                follow_time=np.uint32(0),
                split_time=np.uint32(0),
                data_length=np.uint32(0),
                logging_file_list=[],
                meta_data={"data": {}, "issue": ""}
            )
        else:
            self.message_counter += 1
            config_msg = DataRecordConfigMsg(
                header=Header(time.time_ns(), 21, self.message_counter, 0),
                logging_directory_path="",
                logging_mode=np.uint32(0),
                history_time=np.uint32(0),
                follow_time=np.uint32(0),
                split_time=np.uint32(0),
                data_length=np.uint32(0),
                logging_file_list=[],
                meta_data={"data": {}, "issue": ""}
            )
            
        # Calculate body length
        body_size = message_handler.calculate_body_size()
        config_msg.header.body_length = np.uint32(body_size)
            
        serialized_data = message_handler.make_package(config_msg)
        
        for backend in self.backends:
            if backend["ready"]:
                backend["sockets"][1].sendall(serialized_data)
    
    def toggle_action(self):
        if not self.is_toggle_on:  # Sending START
            success, failed_backends = self.send_tcp_message()
            if success:
                self.toggle_btn.setText("End")
                self.toggle_btn.setStyleSheet("""
                    QPushButton {
                        font-size: 32px;
                        font-weight: bold;
                        padding: 5px;
                        background-color: #ff9999;
                        color: white;
                        border-radius: 5px;
                    }
                    QPushButton:hover {
                        background-color: #ff8080;
                    }
                """)
                self.is_toggle_on = True
                self.event_btn.setEnabled(False)  # Disable event button after successful START
            else:
                # If START fails, notify other backend
                if len(failed_backends) < len(self.backends):
                    failure_message = f"CONNECTION_FAIL:{','.join(failed_backends)}"
                    self.send_tcp_message(failure_message)
                
                self.toggle_btn.setText("Start")
                self.toggle_btn.setStyleSheet("""
                    QPushButton {
                        font-size: 32px;
                        font-weight: bold;
                        padding: 5px;
                        background-color: #4CAF50;
                        color: white;
                        border-radius: 5px;
                    }
                    QPushButton:hover {
                        background-color: #45a049;
                    }
                """)
                self.is_toggle_on = False
                self.event_btn.setEnabled(True)  # Enable event button if START fails
        else:  # Sending END
            success, _ = self.send_tcp_message(start=False)
            if success:
                self.toggle_btn.setText("Start")
                self.toggle_btn.setStyleSheet("""
                    QPushButton {
                        font-size: 32px;
                        font-weight: bold;
                        padding: 5px;
                        background-color: #4CAF50;
                        color: white;
                        border-radius: 5px;
                    }
                    QPushButton:hover {
                        background-color: #45a049;
                    }
                """)
                self.is_toggle_on = False
                self.event_btn.setEnabled(True)  # Enable event button when END is sent

    def send_event(self):
        # 이벤트 전송 로직 구현 필요
        pass
    
    def enable_event_button(self):
        self.event_btn.setEnabled(True)
        self.timer.stop()

    def closeEvent(self, event):
        # 프로그램 종료 시 모든 소켓 정리
        for backend in self.backends:
            for socket in backend["sockets"]:
                if socket is not None:
                    try:
                        socket.close()
                    except:
                        pass
        event.accept()

if __name__ == "__main__":
    app = QApplication(sys.argv)
    app.setStyle('Fusion')
    window = ControlApp()
    window.show()
    sys.exit(app.exec_()) 